programs/mpl-core/src/plugins/verified_creators.rs

```diff
@@ -0,0 +1,205 @@
+) -> Result<SignatureChangeIndices, ProgramError> {
+    let mut existing_map: BTreeMap<Pubkey, &VerifiedCreatorsSignature> = BTreeMap::new();
+    if let Some(verified_creators) = verified_creators {
+        for sig in verified_creators.signatures.iter() {
+            existing_map.insert(sig.address, sig);
+        }
+    }
+
+    let new_signatures: HashSet<_> = new_verified_creators
+        .signatures
+        .iter()
+        .map(|sig| sig.address)
+        .collect();
+
+    if new_verified_creators.signatures.len() != new_signatures.len() {
+        // Ensure there are no duplicate signatures
+        solana_program::msg!("Verified creators: Rejected");
+        return Err(MplCoreError::InvalidPluginSetting.into());
+    }
+
+    let mut result = SignatureChangeIndices {
+        added: Vec::new(),
+        changed: Vec::new(),
+        removed: Vec::new(),
+    };
+
+    for (i, sig) in new_verified_creators.signatures.iter().enumerate() {
+        match existing_map.get(&sig.address) {
+            Some(existing_sig) => {
+                if existing_sig.verified != sig.verified {
+                    result.changed.push(i as u8);
+                }
+            }
+            None => {
+                result.added.push(i as u8);
+            }
+        }
+    }
+
+    if let Some(verified_creators) = verified_creators {
+        for (i, sig) in verified_creators.signatures.iter().enumerate() {
+            if !new_signatures.contains(&sig.address) {
+                result.removed.push(i as u8);
+            }
+        }
+    }
+
+    Ok(result)
+}
+
+fn validate_verified_creators_as_creator(
+    new_verified_creators: &VerifiedCreators,
+    verified_creators: &VerifiedCreators,
+    authority: &Pubkey,
+) -> Result<ValidationResult, ProgramError> {
+    // Track any changes in verification status
+    let changes = calculate_signature_changes(new_verified_creators, Some(verified_creators))?;
+
+    if !changes.added.is_empty() || !changes.removed.is_empty() {
+        // creators cannot add new allowable signatures or remove existing ones
+        solana_program::msg!("Verified creators: Rejected");
+        return Err(MplCoreError::MissingSigner.into());
+    }
+
+    for change in changes.changed.iter() {
+        let sig = &new_verified_creators.signatures[*change as usize];
+        if &sig.address != authority {
+            // creators may only change their own verified status
+            solana_program::msg!("Verified creators: Rejected");
+            return Err(MplCoreError::MissingSigner.into());
+        }
+    }
+
+    Ok(ValidationResult::Approved)
+}
+
+fn validate_verified_creators_as_plugin_authority(
+    new_verified_creators: &VerifiedCreators,
+    verified_creators: Option<&VerifiedCreators>,
+    authority: &Pubkey,
+) -> Result<ValidationResult, ProgramError> {
+    // Plugin auth is allowed to change anything except the verified status of a signature to true
```

Comment:
The legacy behavior in TM and Bubblegum includes:
* The update authority can add/remove unverified creators (supported in this PR by plugin authority).
* The update authority can add itself as verified or set itself to verified because it is a signer (supported in this PR).
* The update authority cannot remove or unverify any existing verified creators, other than itself because it is a signer (this behavior is not currently supported).

I suspect we want to reflect this legacy behavior.

Reply:
good point